### PR TITLE
Add pure operations (`pure_map`, `pure_flat_map`) that use memoization to improve performance

### PR DESCRIPTION
**Pull request by samuelchassot**
# Description

Propose versions of `map` and `flat_map` that use memoization to improve performance. This works only for pure functions so we add a new function instead of modifying `map`. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by samuelchassot** [comment]
@Gondolav WDYT? I made additions only to `list` for now to open the discussion.

**Comment by Gondolav** [comment]
I like it, but you could use directly `functools.cache()` on the function to apply ;)

**Comment by samuelchassot** [comment]
haha thanks, didn't know this package :)

**Review by Gondolav — CHANGES_REQUESTED** [review]
That's great! A few remarks:

- You could use comprehensions with f_cached stored before the comprehension for most methods. Also, pay attention when you can use generators instead of materialised collections, they are more performant for our use case.
- You can do the same with `filter` and `filter_not` I think, so that we mirror what we did with `par` functions.

File: pyfuncol/dict.py [hunk 1]
```diff
@@ -291,6 +292,55 @@ def par_map(self: Dict[A, B], f: Callable[[Tuple[A, B]], Tuple[C, D]]) -> Dict[C
...
+
+def pure_flat_map(
+    self: Dict[A, B], f: Callable[[Tuple[A, B]], Dict[C, D]]
+) -> Dict[C, D]:
+    """
+    Builds a new dict by applying a function to all elements of this dict and using the elements of the resulting collections using memoization to improve performance.
+
+    WARNING: f must be a PURE function i.e., calling f on the same input must always lead to the same result!
+
+    type A must be hashable using `hash()` function.
```
Comment (Gondolav) [inline pyfuncol/dict.py:328]:
```suggestion
    Type A must be hashable using `hash()` function.
```

File: pyfuncol/list.py [hunk 2]
```diff
@@ -330,6 +331,49 @@ def par_flat_map(self: List[A], f: Callable[[A], List[B]]) -> List[B]:
     return [z for y in applications for z in y]
 
 
+def pure_map(self: List[A], f: Callable[[A], B]) -> List[B]:
+    """
+    Builds a new list by applying a function to all elements of this list using memoization to improve performance.
+
+    WARNING: f must be a PURE function i.e., calling f on the same input must always lead to the same result!
+
+    type A must be hashable using `hash()` function.
```
Comment (Gondolav) [inline pyfuncol/list.py:340]:
```suggestion
    Type A must be hashable using `hash()` function.
```

File: pyfuncol/list.py [hunk 3]
```diff
@@ -330,6 +331,49 @@ def par_flat_map(self: List[A], f: Callable[[A], List[B]]) -> List[B]:
...
+    return res
+
+
+def pure_flat_map(self: List[A], f: Callable[[A], List[B]]) -> List[B]:
+    """
+    Builds a new list by applying a function to all elements of this list and using the elements of the resulting collections using memoization to improve performance..
+
+    WARNING: f must be a PURE function i.e., calling f on the same input must always lead to the same result!
+
+    type A must be hashable using `hash()` function.
```
Comment (Gondolav) [inline pyfuncol/list.py:361]:
```suggestion
    Type A must be hashable using `hash()` function.
```

File: pyfuncol/set.py [hunk 4]
```diff
@@ -252,6 +253,52 @@ def par_flat_map(self: Set[A], f: Callable[[A], Set[B]]) -> Set[B]:
...
+# Pure operations
+
+
+def pure_map(self: Set[A], f: Callable[[A], B]) -> Set[B]:
+    """
+    Builds a new set by applying a function to all elements of this set using memoization to improve performance.
+
+    WARNING: f must be a PURE function i.e., calling f on the same input must always lead to the same result!
+
+    type A must be hashable using `hash()` function.
```
Comment (Gondolav) [inline pyfuncol/set.py:265]:
```suggestion
    Type A must be hashable using `hash()` function.
```

File: pyfuncol/set.py [hunk 5]
```diff
@@ -252,6 +253,52 @@ def par_flat_map(self: Set[A], f: Callable[[A], Set[B]]) -> Set[B]:
...
+    return res
+
+
+def pure_flat_map(self: Set[A], f: Callable[[A], Set[B]]) -> Set[B]:
+    """
+    Builds a new set by applying a function to all elements of this set and using the elements of the resulting collections using memoization to improve performance.
+
+    WARNING: f must be a PURE function i.e., calling f on the same input must always lead to the same result!
+
+    type A must be hashable using `hash()` function.
```
Comment (Gondolav) [inline pyfuncol/set.py:286]:
```suggestion
    Type A must be hashable using `hash()` function.
```

**Review by Gondolav — CHANGES_REQUESTED** [review]

**Comment by samuelchassot** [comment]
I did the modifications :) @Gondolav

**Review by Gondolav — CHANGES_REQUESTED** [review]
Ok perfect, maybe just update the example to use `pure_filter` and `pure_filter_not` and we can merge.

File: README.md [hunk 6]
```diff
@@ -56,6 +56,20 @@ import pyfuncol
 # {"a": 1, "b": 4, "c": 9}
 ```
 
+pyfuncol provides operations leveraging memoization to improve performance (for now `pure_map` and `pure_flat_map`). BEWARE: these versions work only
```
Comment (Gondolav) [inline README.md:59]: Actually, could you update the README to do the same as here but for parallel operations? So we have 3 blocks of code, one for the normal ops, one for the parallel ones and one for the pure

Comment (samuelchassot) [reply README.md:72]: Yep, sure :)

**Comment by samuelchassot** [comment]
done 👍

**Review by Gondolav — APPROVED** [review]

**Pull request merged**